### PR TITLE
Dynamic loading of crypto library and support for OpenSSL 1.0.2

### PR DESCRIPTION
**Pull request by enasser**
In this changeset, the following changes are added:
    - Support of dynamic loading of OpenSSL crypto library
    - Support for 1.0.2, 1.1.0 and 1.1.1 OpenSSL releases

Signed-off-by: Nasser Ebrahim <enasser@in.ibm.com>

### CODE REVIEW

**Comment by enasser** [comment]
openssl 1.0.2 build will fail while compile with gnu make 7.3 as it specify dynamically generating <arch>cpuid.o in the Makefile. The issue is resolved in openssl 1.1.x. As OpenJDK11 uses gnu make 7.3 for compilation, make sure openssl is separately compiled with gnu make 4.1 while using --with-openssl=fetched  option is used.

**Comment by enasser** [comment]
As same function is used to retrieve all the symbols of crypto library with different data types,  compiler issue a warning while assigning the pointers. In gnu make 7.3, the warning are considered as error by default and fail. Use --disable-warnings-as-errors compile option to avoid the failure.

**Comment by enasser** [comment]
@andrew-m-leonard @pshipton  Please review and merge. I have built and tested on Windows, x64, z64 and le64 linux.. AIX and Mac build is in progress but am not expecting any failures.

**Comment by pshipton** [comment]
jenkins compile xlinux,plinux,zlinux,win

**Comment by pshipton** [comment]
jenkins compile win

**Comment by pshipton** [comment]
@enasser xlinux compile failed
https://ci.eclipse.org/openj9/job/PullRequest-Compile-JDK11-linux_x86-64_cmprssptrs-OpenJDK11/33
```
17:28:57 === Output from failing command(s) repeated here ===
17:28:57 * For target support_native_java.base_libjncrypto_NativeCrypto.o:
17:28:57 /home/jenkins/workspace/PullRequest-Compile-JDK11-linux_x86-64_cmprssptrs-OpenJDK11/closed/src/java.base/share/native/libjncrypto/NativeCrypto.c: In function ‘Java_jdk_crypto_jniprovider_NativeCrypto_loadCrypto’:
17:28:57 /home/jenkins/workspace/PullRequest-Compile-JDK11-linux_x86-64_cmprssptrs-OpenJDK11/closed/src/java.base/share/native/libjncrypto/NativeCrypto.c:136:18: error: assignment from incompatible pointer type [-Werror=incompatible-pointer-types]
17:28:57      OSSL_version = (char * (*))find_crypto_symbol(handle, "OpenSSL_version");
17:28:57                   ^
17:28:57 /home/jenkins/workspace/PullRequest-Compile-JDK11-linux_x86-64_cmprssptrs-OpenJDK11/closed/src/java.base/share/native/libjncrypto/NativeCrypto.c:139:22: error: assignment from incompatible pointer type [-Werror=incompatible-pointer-types]
17:28:57          OSSL_version = (char *)find_crypto_symbol(handle, "SSLeay_version");
17:28:57                       ^
17:28:57 /home/jenkins/workspace/PullRequest-Compile-JDK11-linux_x86-64_cmprssptrs-OpenJDK11/closed/src/java.base/share/native/libjncrypto/NativeCrypto.c:175:15: error: assignment from incompatible pointer type [-Werror=incompatible-pointer-types]
17:28:57      OSSL_sha1 = (const EVP_MD* *)find_crypto_symbol(handle, "EVP_sha1");
17:28:57                ^
17:28:57 /home/jenkins/workspace/PullRequest-Compile-JDK11-linux_x86-64_cmprssptrs-OpenJDK11/closed/src/java.base/share/native/libjncrypto/NativeCrypto.c:176:17: error: assignment from incompatible pointer type [-Werror=incompatible-pointer-types]
17:28:57      OSSL_sha256 = (const EVP_MD* *)find_crypto_symbol(handle, "EVP_sha256");
17:28:57    ... (rest of output omitted)
```

**Comment by andrew-m-leonard** [comment]
@enasser Various compile errors still, are you able to look at these?

**Comment by enasser** [comment]
@andrew-m-leonard @pshipton Regarding the compile errors, I have already explained that in my previous comment. 

> As same function is used to retrieve all the symbols of crypto library with different data types, compiler issue a warning while assigning the pointers. In gnu make 7.3, the warning are considered as error by default and fail. Use **--disable-warnings-as-errors** compile option to avoid the failure.

**Comment by DanHeidinga** [comment]
We should not be introducing new warnings that require compiling with `--disable-warnings-as-errors`.  These appear to be some incorrect casts for the function pointers which should be fixable.

Some typedefs may make this easier to read / write:
```c
typedef const char* ssl_version_t(int);
.....
OSSL_version = (ssl_version_t*)find_crypto_symbol(handle, "OpenSSL_version");
.....
OSSL_version = (ssl_version_t *)find_crypto_symbol(handle, "SSLeay_version");
```
(Above code has not be compiled)

**Comment by pshipton** [comment]
@enasser Using `--disable-warnings-as-errors` is not desirable or feasible. It would require many changes across OpenJ9 and Adopt builds, and possibly vendor's internal builds as well (i.e. IBM's internal builds). We also want to keep the code warning free. Turning on this option allows not just this code containing warnings but future code as well, making it harder to recover and remove `--disable-warnings-as-errors` at a later time.

This code should not be merged until the warnings are fixed.

**Comment by ddoliver** [comment]
@enasser any updates on this, ETA please?

**Comment by enasser** [comment]
@pshipton @DanHeidinga Thank you for your inputs and review comments. Am currently working to fix the compile warning.

@ddoliver I will make it ready by tomorrow.

**Comment by enasser** [comment]
@pshipton  I have fixed the compile errors, squash and push. Please review and merge.

**Comment by pshipton** [comment]
jenkins compile zlinux,osx,win

**Comment by pshipton** [comment]
jenkins compile win

**Comment by pshipton** [comment]
jenkins test sanity win

**Comment by pshipton** [comment]
@jdekonin is this a machine problem, did something change?
osx1013-x86-4
https://ci.eclipse.org/openj9/job/PullRequest-Compile-JDK11-osx_x86-64_cmprssptrs-OpenJDK11/6
The failure doesn't seem to have anything to do with this change.
```
16:55:30 In file included from /Users/jenkins/workspace/PullRequest-Compile-JDK11-osx_x86-64_cmprssptrs-OpenJDK11/src/java.desktop/macosx/native/libawt_lwawt/awt/AWTSurfaceLayers.h:26:
16:55:30 /Users/jenkins/workspace/PullRequest-Compile-JDK11-osx_x86-64_cmprssptrs-OpenJDK11/src/java.desktop/unix/native/include/jawt_md.h:29:10: fatal error: 'X11/Xlib.h' file not found
16:55:30 #include <X11/Xlib.h>
```

**Comment by pshipton** [comment]
jenkins compile osx

**Comment by pshipton** [comment]
jenkins test sanity plinux

**Comment by jdekonin** [comment]
To my knowledge, nothing has changed in the OSX machines in the past week.

**Comment by pshipton** [comment]
@enasser Windows failed, I think because of 
```
17:32:02 make[3]: *** No rule to make target '/cygdrive/c/OpenSSL-1.1.1-x86_64/libeay32.dll', needed by '/cygdrive/c/Users/jenkins/workspace/PullRequest-Sanity-JDK11-win_x86-64_cmprssptrs-OpenJDK11/build/windows-x86_64-normal-server-release/support/modules_libs/java.base/libeay32.dll'.  Stop.
17:32:02 make[3]: *** Waiting for unfinished jobs....
17:32:02 make[2]: *** [make/Main.gmk:162: java.base-copy] Error 2
```

**Comment by pshipton** [comment]
jenkins tests sanity osx

**Comment by pshipton** [comment]
jenkins test sanity osx

**Comment by pshipton** [comment]
I don't know why osx is failing to build. It has failed on two different machines. I'll try a build outside of this PR to confirm the changes in this PR are causing the build failure.

https://ci.eclipse.org/openj9/job/PullRequest-Compile-JDK11-osx_x86-64_cmprssptrs-OpenJDK11/7/
https://ci.eclipse.org/openj9/job/PullRequest-Sanity-JDK11-osx_x86-64_cmprssptrs-OpenJDK11/4/

**Comment by pshipton** [comment]
I do notice this output in the builds.
`17:26:34 closed/get_openssl_source.sh: line 63: [1.1.: command not found`

**Comment by enasser** [comment]
Both the failures are because of missing X11 library in the system.
OpenJDK11/src/java.desktop/unix/native/include/jawt_md.h:29:10: fatal error: 'X11/Xlib.h' file not found
https://ci.eclipse.org/openj9/job/PullRequest-Compile-JDK11-osx_x86-64_cmprssptrs-OpenJDK11/7/consoleFull
https://ci.eclipse.org/openj9/job/PullRequest-Sanity-JDK11-osx_x86-64_cmprssptrs-OpenJDK11/4/consoleFull

The passing build run on osx1013-x86-3 where as the failing compile job run on osx1013-x86-4. Am not sure the machine at which the failing sanity job run. Please run the pull job in osx1013-x86-3 and see whether it resolves the issue.

**Comment by pshipton** [comment]
@enasser the failing sanity job ran on osx1013-x86-3, the same machine that passed outside of this PR. Although the failures are related to `fatal error: 'X11/Xlib.h'`, something in this change seem to be causing it.

**Comment by pshipton** [comment]
@enasser there is also the Windows failure https://github.com/ibmruntimes/openj9-openjdk-jdk11/pull/90#issuecomment-455368729

File: closed/src/java.base/share/native/libjncrypto/NativeCrypto.c [hunk 1]
```diff
@@ -26,13 +26,78 @@
...
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
-#include <assert.h>
 
 #include "jdk_crypto_jniprovider_NativeCrypto.h"
+#include "NativeCrypto_md.h"
+
+//Type definitions of function pointers
+typedef char * OSSL_error_string_t(unsigned long, char *);
```
Comment (keithc-ca) [inline closed/src/java.base/share/native/libjncrypto/NativeCrypto.c]: These are technically, *not* function pointers, if you want a function pointer you should write:
> typedef char * (*OSSL_error_string_t)(unsigned long, char *);

and replace `OSSL_error_string_t *` with `OSSL_error_string_t`.

File: closed/src/java.base/share/native/libjncrypto/NativeCrypto.c [hunk 2]
```diff
@@ -543,58 +741,51 @@ JNIEXPORT jint JNICALL Java_jdk_crypto_jniprovider_NativeCrypto_GCMDecrypt
...
-            evp_gcm_cipher = EVP_aes_192_gcm();
+            evp_gcm_cipher = (*OSSL_aes_192_gcm)();
             break;
         case 32:
-            evp_gcm_cipher = EVP_aes_256_gcm();
+            evp_gcm_cipher = (*OSSL_aes_256_gcm)();
             break;
         default:
-            break;
+            assert(32);
```
Comment (keithc-ca) [inline closed/src/java.base/share/native/libjncrypto/NativeCrypto.c]: I don't think `assert()` is the best way to handle errors; and this use is just wrong as `assert()` only treats a zero argument as a failure.

File: closed/src/java.base/unix/native/libjncrypto/NativeCrypto_md.c [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+/* Find the symbol in the crypto library (return NULL if not found) */
+void * find_crypto_symbol(void *handle, const char *symname) {
+    void * symptr;
+
+    symptr =  dlsym(handle, symname);
+
+    if (symptr == NULL) {
+        fputs (dlerror(), stderr);
```
Comment (keithc-ca) [inline closed/src/java.base/unix/native/libjncrypto/NativeCrypto_md.c]: This error handling seems inappropriate here.

File: closed/src/java.base/windows/native/libjncrypto/NativeCrypto_md.c [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+/* Find the symbol in the crypto library (return NULL if not found) */
+void * find_crypto_symbol(void *handle, const char *symname) {
+    void * symptr;
+
+    symptr =  GetProcAddress(handle, symname);
+
+    return symptr;
+}
```
Comment (keithc-ca) [inline closed/src/java.base/windows/native/libjncrypto/NativeCrypto_md.c]: There should be a final end-of-line for every text file.

**Comment by davehobbs** [comment]
jenkins compile xlinux

**Comment by andrew-m-leonard** [comment]
jenkins compile plinux

**Comment by keithc-ca** [comment]
Why the jenkins builds? I don't see responses to any of my comments.

**Comment by davehobbs** [comment]
Just to see if I could run Jenkins builds - which I can't, and Andrew can.

**Comment by groeges** [comment]
Superceeded by pull request https://github.com/ibmruntimes/openj9-openjdk-jdk11/pull/135

**Pull request closed without merging**